src/notes/components/Toolbar.tsx

```diff
@@ -48,25 +50,42 @@ const titles = {
 interface ToolbarProps {
   os?: "mac" | "other"
   note?: Note
+  theme?: Theme
 }
 
-const Toolbar = ({ os, note }: ToolbarProps): h.JSX.Element => {
+const Toolbar = ({ os, note, theme }: ToolbarProps): h.JSX.Element => {
   const getTitle = useCallback((key: string) => {
     const title = (os && titles[key] && titles[key][os]) || "";
     return title;
   }, [os]);
+
   const [submenu, setSubmenu] = useState<string | null>(null);
 
-  const toggleSubmenu = (event: MouseEvent): void => {
+  const toggleSubmenu = useCallback((event: MouseEvent): void => {
     if (event.target === event.currentTarget) {
+      range.save();
       const submenuToToggle = (event.target as HTMLDivElement).id;
       setSubmenu((prev) => prev === submenuToToggle ? null : submenuToToggle);
     }
-  };
+  }, []);
+
+  useEffect(() => {
+    range.restore();
+  }, [submenu]);
 
   const [insertImageModalProps, setInsertImageModalProps] = useState<InsertImageModalProps | null>(null);
   const [insertLinkModalProps, setInsertLinkModalProps] = useState<InsertLinkModalProps | null>(null);
 
+  const ThemeColorFilter = useCallback((color: string): boolean => {
+    if (
+      (theme === "dark" && ["black", "white", "silver", "purple"].includes(color)) ||
+      (theme === "light" && ["white", "purple"].includes(color))
+    ) {
+      return false;
+    }
+    return true;
+  }, [theme]);
```

Comment:
Filter colors based on the Theme. White color text is not needed in Light theme. Same as Black color text is not needed in Dark Theme.